File: App/Resources/AlgorithmDetails/weavesortrecursive/weavesortrecursive.kt
fun sort(arr: Array<Int>) {
  val end = arr.size

  fun compSwap(a: Int, b: Int) {
    if (b < end && arr[a] > arr[b]) {
      val temp = arr[a]
      arr[a] = arr[b]
      arr[b] = temp
    }
  }

  fun circle(pos: Int, ln: Int, gap: Int) {
    if (ln < 2) return
    var i = 0
    while (2 * i < (ln - 1) * gap) {
      compSwap(pos + i, pos + (ln - 1) * gap - i)
      i += gap
    }
    circle(pos, ln / 2, gap)
    if (pos + ln * gap / 2 < end) {
      circle(pos + ln * gap / 2, ln / 2, gap)
    }
  }

  fun weaveCircle(pos: Int, ln: Int, gap: Int) {
    if (ln < 2) return
    weaveCircle(pos, ln / 2, 2 * gap)
    weaveCircle(pos + gap, ln / 2, 2 * gap)
    circle(pos, ln, gap)
  }

  var padded = 1
  while (padded < end) {
    padded *= 2
  }

  weaveCircle(0, padded, 1)
}

fun main() {
  var array = arrayOf<Int>(
    0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56,
  )
  sort(array)
  println("[%s]".format(array.joinToString(", ")))
}

File: App/Resources/AlgorithmDetails/weavesortrecursive/weavesortrecursive.go
package main

import (
	"fmt"
)

func compSwap(arr []int, a, b, end int) {
	if b < end && arr[a] > arr[b] {
		arr[a], arr[b] = arr[b], arr[a]
	}
}

func circle(arr []int, pos, ln, gap, end int) {
	if ln < 2 {
		return
	}
	i := 0
	for 2*i < (ln-1)*gap {
		compSwap(arr, pos+i, pos+(ln-1)*gap-i, end)
		i += gap
	}
	circle(arr, pos, ln/2, gap, end)
	if pos+ln*gap/2 < end {
		circle(arr, pos+ln*gap/2, ln/2, gap, end)
	}
}

func weaveCircle(arr []int, pos, ln, gap, end int) {
	if ln < 2 {
		return
	}
	weaveCircle(arr, pos, ln/2, 2*gap, end)
	weaveCircle(arr, pos+gap, ln/2, 2*gap, end)
	circle(arr, pos, ln, gap, end)
}

func sort(arr []int) []int {
	end := len(arr)
	padded := 1
	for padded < end {
		padded *= 2
	}
	weaveCircle(arr, 0, padded, 1, end)
	return arr
}

func main() {
	array := []int{0, 39, 21, 62, 91, 77, 14, 23,
		90, 69, 51, 81, 68, 83, 32, 56}
	fmt.Println(sort(array))
}
